README
POE/Component/Daemon version 0.1001
=================================


This component attempts to make building daemons in POE as easy as possible.
It implements pre-forking, post-forking and non-forking daemons.

INSTALLATION

To install this module type the following:

   perl Makefile.PL
   make
   make test
   make install

DEPENDENCIES

This module requires these other modules and libraries:

    POE
    POE::API::Peek

COPYRIGHT AND LICENCE

Put the correct copyright and licence information here.

Copyright (C) 2004,2005,2006,2009,2010 Philip Gwyn

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself. 



Changes
Revision history for Perl extension POE::Component::Daemon.

0.1300 2011/May/18
    - Removed dependence on POE::API::Peek
    - Fixed, tested and working on POE 1.311

0.1204 2011/Fev/18
    - Moved all ->yield( 'shutdown' ) to ->do_shutdown();
    - daemon_shutdown is now expedited with ->call()
    - Tweak cookie crumbs

0.1203 2011/Jan/06
    - Check for a working shmget() in Makefile.PL

0.1201 2010/Dec/13
    - Fixed test cases!
    - Stronger exit code for detach

0.1200 2010/Dec/13
    - shutdown is also a signal
    - only remove alias once
    - tweak verbose cookie crumbs
    - moved clutter into eg/

0.1100 2009/Sep/16
    - Use Port=>0 so that the OS assigns us a port number.  This way
        maybe we won't always fail on darwin.
    - use POE::Kernel->sig_child()

0.1008 2009/May/6
    - Make CPANTS happier

0.1007 2008/Nov/25
    - Added a call to poe_kernel->_data_sig_initialize to work around a bug
        in POE 1.003
    - A transition req->long or long->req no longer counted a request

0.1006 2007/Aug/01
    - Added daemon_HUP

0.1004 2007/Jun/29
    - Clean up dupped sig_TERM

0.1003 2006/Nov/15
    - Use $^X in test cases

0.1002 2006/Nov/09
    - Reconciled with SVN

0.1001 2006/Sep/14
    - Remove signal handlers so that kernel can still exit.  This is 
      required in the new, refcount on sig kernels (post 0.36)
      Thank you Andrew Hoying

0.10 2006/Feb/22
    - First public release

0.01  Wed Apr  7 15:38:51 2004
	- original version; created by h2xs 1.22 with options
		-AX -b 5.4.5 POE::Component::Daemon



Makefile.PL
#!/usr/bin/perl
use strict;
use warnings;
use 5.00405;
use ExtUtils::MakeMaker;
# See lib/ExtUtils/MakeMaker.pm for details of how to influence
# the contents of the Makefile that is written.

eval "use ExtUtils::MakeMaker::Coverage;";

# Make sure shmget works properly
eval {
    require "IPC/SysV.pm";
    my $mem = shmget(IPC::SysV::IPC_PRIVATE(), 32, IPC::SysV::S_IRWXU())
                        ||
              shmget(IPC::SysV::IPC_PRIVATE(), 32, IPC::SysV::S_IRWXU());
    $mem or die "shmget failed: $!";
    if( $mem ) {
        shmctl( $mem, IPC::SysV::IPC_RMID(), 0 )
            or die "shmctl failed: $!";
    }
};
if( $@ ) {
    warn "Shared memory not available on $^O: $@\n";
    exit 0;
}

WriteMakefile(
    'AUTHOR'         => 'Philip Gwyn <gwyn -at- cpan.org>',
    'NAME'	    	=> 'POE::Component::Daemon',
    'VERSION_FROM'	=> 'Daemon.pm', # finds $VERSION
    'LICENSE'       => 'perl',
    'PREREQ_PM'		=> {
        'IPC::SysV'        => 1,
        'POE'              => 1,
        'File::Basename'   => 1,
        'File::Copy'       => 1,
        # 'POE::API::Peek'  => 1        
    },
    dist=>{COMPRESS=>'gzip -9f', EXT=>'gz'},
);


Daemon/Scoreboard.pm
#$Id: Scoreboard.pm 700 2010-12-15 16:43:10Z fil $
########################################################
package POE::Component::Daemon::Scoreboard;

use 5.00405;
use strict;

use vars qw($VERSION $UNIQUE);

use IPC::SysV qw(IPC_PRIVATE S_IRWXU IPC_CREAT SEM_UNDO);
use Carp;

$VERSION = '0.1100';

sub DEBUG () { 0 }

########################################################
sub new
{
    my($package, $N)=@_;
    if($UNIQUE) {
        warn "This should be only one $package.  Reusing previous one.";
        return $UNIQUE;
    }

    my $self=bless {N=>$N}, $package;

    # On linux, 2.6 kernels (at least), the first call after a reboot will
    # fail, second and subsequent will succeed.
    $self->{mem} = shmget(IPC_PRIVATE, $N, S_IRWXU) 
                        || 
                   shmget(IPC_PRIVATE, $N, S_IRWXU);
    die "$$: Unable to create shared memory: $!\n" unless $self->{mem};

    $self->{slots}=[reverse 0..($N-1)];

    my $blank=' ' x $N;
    shmwrite($self->{mem}, $blank, 0, $N);

    $UNIQUE=$self;

    return $self;
}

########################################################
sub read_all
{
    my($self)=@_;

    my $str=" " x $self->{N};
    shmread($self->{mem}, $str, 0, $self->{N})
        or die "Unable to read shared memory: $!\n";

    my $ret=[split //, $str];
    return $ret;
}

########################################################
sub add
{
    my($self, $value)=@_;
    return unless @{$self->{slots}};
    my $slot=pop @{$self->{slots}};
    DEBUG and warn "Adding slot $slot";
    $self->write($slot, $value);
    return $slot;
}

########################################################
sub drop
{
    my($self, $slot)=@_;
    if($slot >= $self->{N}) {
        carp "$slot isn't a known slot\n";
        return;
    }
    $self->write($slot, '.');
    DEBUG and warn "Dropped slot $slot";
    push @{$self->{slots}}, $slot;
    return;
}

########################################################
sub write
{
    my($self, $slot, $value)=@_;
    if($slot >= $self->{N}) {
        carp "$slot isn't a known slot\n";
        return;
    }

    $value=substr($value, 0, 1);
    DEBUG and warn "Setting slot $slot to $value";

    shmwrite($self->{mem}, $value, $slot, 1)
        or warn "Writing shared memory slot $slot: $!";

    return;
}

########################################################
sub read
{
    my($self, $slot)=@_;
    return unless defined $slot;

    if($slot >= $self->{N}) {
        carp "$slot isn't a known slot\n";
        return;
    }
    DEBUG and warn "Reading value $slot";

    my $value=" ";
    shmread($self->{mem}, $value, $slot, 1)
        or warn "Reading shared memory slot $slot: $!";
    return $value;
}

########################################################
sub status
{
    my($self)=@_;
    my @ret;

    my $n=$self->read_all();
    push @ret, ref($self);
    push @ret, "$self->{N} slots in the scoreboard";
    push @ret, join '', "Slots [", @$n, "]";
    push @ret, (0+@{$self->{slots}})." slots free";

    return join "\n    ", @ret;
}

1;

__DATA__



Daemon/ScoreboardSemaphore.pm
#$Id: ScoreboardSemaphore.pm 700 2010-12-15 16:43:10Z fil $
########################################################
package POE::Component::Daemon::Scoreboard;

use 5.00405;
use strict;

use vars qw($VERSION $UNIQUE);

use IPC::SysV qw(IPC_PRIVATE S_IRWXU IPC_CREAT SEM_UNDO);
use IPC::Semaphore;
use Carp;

$VERSION = '0.1100';

sub DEBUG () { 1 }

sub FS () { "\x1c" }
sub EOT () { "\x03" }

########################################################
sub new
{
    my($package, $N)=@_;
    if($UNIQUE) {
        warn "This should be only one $package.  Reusing previous one.";
        return $UNIQUE;
    }

    my $self=bless {N=>$N}, $package;

    $self->{semaphore}=IPC::Semaphore->new(IPC_PRIVATE, 2, 
                                                S_IRWXU | IPC_CREAT);
    $self->{semaphore}->setall(1, 1);
    $self->{have}='';

    my $size = $N* 5 + 1;
    $self->{mem}=shmget(IPC_PRIVATE, $size, S_IRWXU);
    $self->{mem_size}=$size;
    $self->{slots}=[reverse 0..($N-1)];

    my $blank=EOT;
    shmwrite($self->{mem}, $blank, 1, $self->{mem});

    die "Unable to create shared memory: $!\n" unless $self->{mem};

    $UNIQUE=$self;

    return $self;
}

########################################################
sub grab_sem
{
    my($self)=@_;
    return if $self->{have};
    $self->{semaphore}->op(0, -1, SEM_UNDO) or die "Decrementing semaphore: $!\n";
    $self->{have}=1;
    return;
}

########################################################
sub release_sem
{
    my($self)=@_;
    return unless $self->{have};
    $self->{semaphore}->op(0, +1, SEM_UNDO) or die "Releasing semaphore: $!\n";
    $self->{have}=0;
    return;
}

########################################################
sub read_all
{
    my($self)=@_;
    $self->grab_sem();

    my $str;
    shmread($self->{mem}, $str, 0, $self->{mem_size})
        or die "Unable to read shared memory: $!\n";
    # truncate it
    substr($str, index($str, EOT))='';

    my $ret=[split FS, $str];

    $self->release_sem;
    return $ret;
}

########################################################
sub write_all
{
    my($self, $values)=@_;
    $self->grab_sem();

    my $str=join FS, @$values;
    $str.=EOT;
    my $length=length($str);

    die "Can't write more then $self->{size} bytes to shared memory!"
        unless $length < $self->{mem_size};

    # pad it
    # $str.=join '', $str, EOT x ($self->{mem_size} - $length);
    shmwrite($self->{mem}, $str, 0, length($str))
        or die "Unable to write shared memory: $!\n";

    $self->release_sem;
    return;
}

########################################################
sub add
{
    my($self, $value)=@_;
    return unless @{$self->{slots}};
    my $slot=pop @{$self->{slots}};
    $self->write($slot, $value);
    DEBUG and warn "Added slot $slot";
    return $slot;
}

########################################################
sub drop
{
    my($self, $slot)=@_;
    if($slot >= $self->{N}) {
        carp "$slot isn't a known slot\n";
        return;
    }
    $self->write($slot, '.');
    DEBUG and warn "Dropped slot $slot";
    push @{$self->{slots}}, $slot;
    return;
}

########################################################
sub write
{
    my($self, $slot, $value)=@_;
    if($slot >= $self->{N}) {
        carp "$slot isn't a known slot\n";
        return;
    }

    $value=substr($value, 0, 4);
    DEBUG and warn "Setting slot $slot to $value";

    $self->grab_sem();
    my $values=$self->read_all();
    $values->[$slot]=$value;

    $self->write_all($values);

    $self->release_sem();
    return;
}

########################################################
sub read
{
    my($self, $slot)=@_;
    if($slot >= $self->{N}) {
        carp "$slot isn't a known slot\n";
        return;
    }
    DEBUG and warn "Reading value $slot";
    my $values=$self->read_all();
    return $values->[$slot];
}

########################################################
sub status
{
    my($self)=@_;
    my @ret;

    push @ret, "$self->{N} slots in the scoreboard";    
    push @ret, (0+@{$self->{slots}}), " slots free";
    my $state=$self->{semaphore}->getncnt(0);
    push @ret, "$state processes want the scoreboard";

    return join "\n", @ret;
}

1;

__DATA__


